torch/lib/c10d/NCCLUtils.hpp

```diff
@@ -57,6 +58,7 @@ class NCCLComm {
   NCCLComm() : NCCLComm(nullptr) {}
 
   ~NCCLComm() noexcept {
```

Comment:
Why lock in the destructor?

Are the raw pointer references to this NCCLComm that may cause multiple threads to access it even when it is being destructed? If not, the mutex should not be needed here.

Reply:
the aborted_ may be accessed by multiple threads?

Reply:
I suggested having a mutex here since won't have a memory barrier without the mutex. As an example, a thread could call ncclCommAbort setting aborted_ = true, now if a separate thread calls the destructor, the destructor might not see the latest update to aborted_ without a lock (causing it to abort the communicator again).

Reply:
If two threads can race on destruction then that's a separate problem.

I would assume that access to the NCCLComm is also somehow protected by a lock (access to a map of some sort), or that it's always wrapped in a shared_ptr. If it's wrapped in a shared_ptr, then by definition, there can only be a single thread calling the destructor.

Reply:
My feeling is that we shouldn't rely on how the class is used in order to guarantee safety for internal details of this class.

Also, my previous comment wasn't about two threads racing on destruction (I don't think the destructor can be called from two threads). It was more about the fact that the thread calling the destructor wouldn't see the latest value of `aborted_` (which might be updated by separate thread earlier) without a memory barrier.

Reply:
sounds good to me that we add a lock here to protect aborted_ value is synced up. but agree that we always create a share_ptr<NCCLComm> in this class and will not be accessed by two threads though.

Reply:
@pritamdamania87 @zhaojuanmao I see, I think that makes sense. Can you add a comment explaining this rationale to the destructor where the lock is acquired? This discussion is highly relevant and would prevent future removal of the code (as others might reason similarly).